### ISSUE TITLE
Huge disk space is being used under `/home/<username>/.cache/vscode-cpptools`

### ISSUE THREAD

**Issue by sly1061101** [microsoft/vscode-cpptools#12066]
As the screenshot shows, it's using ~270GB on my system! 

![image](https://github.com/microsoft/vscode-cpptools/assets/18494936/b178ddeb-26c7-496d-a789-cb54e1f0e5dd)

The intelli sense cache size has been limited to 5GB, which is indeed the size for the `ipch` folder:

![image](https://github.com/microsoft/vscode-cpptools/assets/18494936/6694ac23-244d-48ab-bc39-651d210c5d41)

What are the other folders? Is there a way to also limit their sizes?

**Comment by bobbrow** [comment]
Which version of the extension are you using?  There was a bug in 1.19.4 with the cleanup mechanism and we patched it in 1.19.5.  If you update to the latest version it should clean up the folder the next time you open a C or C++ file.

**Comment by bobbrow** [comment]
The hash-named folders contain the database we generate when we index the workspace folder.  Judging by the size of some of these, I wonder if you are hitting an issue with a symlink pointing outside your workspace folder and resulting in the extension indexing way more than intended.  We don't generally see databases that large.  I don't think even Chromium's database gets that big, but it's been a while since I checked.  Are you working on very large codebases?

**Comment by sly1061101** [comment]
Thanks for the quick response. I do work on a fairly large code base, and I have three copies (git worktree) of the code base. So it's kind of as expected?

**Comment by bobbrow** [comment]
It would be expected to have large databases for large codebases, but 90GB still seems like a lot to me.  If you set the `C_Cpp.loggingLevel` setting to `"debug"` and reload your window, you can search for a string like this to see how many files we processed for the database:
```
Discovering files: 31704 file(s) processed
```
That workspace has a 500MB database associated with it so if I extrapolate to 90GB it would assume there are upwards of 5M files being tracked in your largest database.  If your codebase is actually that big, I would say the number looks in line with my expectations, but if it isn't, we might want to look and see if you do have the symlink problem.

**Comment by sly1061101** [comment]
> BTW, OP: Which software are you using in the image of the OP? I use WinDirStat on Windows, but you seem to be on Linux?

@TryerGit  It's on Ubuntu Linux, a system utility named "Disk Usage Analyzer"

**Comment by bobbrow** [comment]
@sly1061101 by any chance is your repo open source?

**Comment by sly1061101** [comment]
> by any chance is your repo open source?

Unfortunately not, it's my employer's code base

> and you don't need to check for symlink issues because you're not at an unexpected 5M files being indexed.

Uh, ok, so the file numbers is not the issue here? Anyelse I could try?

**Comment by bobbrow** [comment]
One thing you can try is to regenerate the database for your workspace and compare the size.  If you'd like to do that without deleting your existing one, you can either:
* Rename the one in the cache folder to something else (please close all instances of VS Code first to ensure that the DB is not accidentally corrupted) and then reopen your workspace.
* In your `c_cpp_properties.json` file, set a location for the following property in your active configuration (ensure the path is different than the default location ~/.cache/vscode-cpptools):
```
{
  "name": "myConfig",
  ...
  "browse": {
    "databaseFilename": "${workspaceFolder}/.vscode/newDatabase.db"
  }
}
```

**Comment by sly1061101** [comment]
Thanks, I renamed the exisiting one to `vscode-cpptools-bak`, and reopened the several workspaces. I can confirm each unique hash corresponds to one workspace I have:

![image](https://github.com/microsoft/vscode-cpptools/assets/18494936/6f4a69bb-4f23-4eca-9724-5b634d2a2af8)

One question - do you know why in the original folder there could be multiple folders for the same hash (just with `-1`, `-2` postfix)?

I also confirmed the file `.browser.VC.db` is using up the significant disk space:

![image](https://github.com/microsoft/vscode-cpptools/assets/18494936/966ffc17-d68c-46bf-a3cc-85500c1473c1)

**Comment by sly1061101** [comment]
Found another strange behavior, so I opened the workspace that corresponds to `3301dddae....` multiple times, and it seems that every time a new folder is being created:

![image](https://github.com/microsoft/vscode-cpptools/assets/18494936/5b086b84-0f63-49ff-90dd-d0643db71db7)

I don't think I observed this before, otherwise the original folder should have way more than 3 folders for this hash... pretty sure I've opened this workspace a lot of times

**Comment by bobbrow** [comment]
Sorry for disappearing on this issue.  We've been dealing with some regressions related to the release.  One of our engineers noticed a strange issue with his database that caused the extension to start indexing from the folder _above_ the workspace folder.  He opened this issue (#12105) and we hope that it's related to what you're seeing.  We'll have a fix out in 1.19.9 that we'd like for you to try when it's available.  @Colengms

**Comment by sean-mcmanus** [comment]
@sly1061101 Is this fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.19.9. Otherwise we might need more repro info.

**Comment by sly1061101** [comment]
Thank you both for the updates. I will update to 1.19.9, remove the cache directory and monitor if it's resolved

**Comment by sean-mcmanus** [comment]
@sly1061101 VS Code itself is sending us those hash values as the last part of the extensionContext.storageUri, see https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/client.ts#L1193 .

**Comment by bobbrow** [comment]
What @sean-mcmanus is suggesting is that the `-#` paths seem to be coming from VS Code and we don't know why they are creating multiple hashes for the same workspace.  Is there something special about the way you open the workspace or do you open it the same way every time?  Do you have multiple instances of VS Code open at the same time?

**Comment by sly1061101** [comment]
> Is there something special about the way you open the workspace or do you open it the same way every time? Do you have multiple instances of VS Code open at the same time?

It's the same way every time. There's no multiple instance, but I did notice that sometimes the `cpptools` process does not exit immidiately after I close VSCode. However, the duplicated folders for the same workspace would be created even if I wait until `cpptools` process has fully stopped.


I waited for one day after clearing the cache directory, and now they seem to have been recreated and still using considerable amout of space:

```
$ du -sh ~/.cache/vscode-cpptools/
59G     /home/liuyis/.cache/vscode-cpptools/

$ du -sh ~/.cache/vscode-cpptools/*
13G     /home/liuyis/.cache/vscode-cpptools/23ba42b1fb5498717cb6b48f49d3d7f2
1.4G    /home/liuyis/.cache/vscode-cpptools/23ba42b1fb5498717cb6b48f49d3d7f2-1
343M    /home/liuyis/.cache/vscode-cpptools/23ba42b1fb5498717cb6b48f49d3d7f2-2
5.9G    /home/liuyis/.cache/vscode-cpptools/27393237d09da9a54bc9f2744eead072
905M    /home/liuyis/.cache/vscode-cpptools/5c9bc4f05e65f0f64af9ec03c59bcd64
37G     /home/liuyis/.cache/vscode-cpptools/5c9bc4f05e65f0f64af9ec03c59bcd64-1
808M    /home/liuyis/.cache/vscode-cpptools/ipch
```

so it's probably a different issue？

**Comment by Colengms** [comment]
Hi @sly1061101 .  Could you provide the contents of your `c_cpp_properties.json` and any VS Code that settings that related to the C/C++ Extension (anything starting with `C_Cpp`) ?  This is something we usually ask for up front.  This issue could possibly be by-design, if you've specified `/` as one of your include paths or browse paths, for example, causing the extension to try to index your entire file system into the database.

I'd also suggest enabling `"C_Cpp.loggingLevel": "Debug"` and monitoring the output in the C/C++ Output channel.  It will list what files as it parse them, and you may spot files that it shouldn't parse, which may lead to identifying the cause.

You can also open the database itself.  It's a sqlite database and can be opened using a sqlite database viewer app.  You might check the file table, to see if it contains information for files that are not related to your workspace.

**Comment by johnothwolo** [comment]
@Colengms I've noticed this same issue because I use vscode to view large C/C++ projects like xnu and FreeBSD. The cumulative space taken up by ipch files is huge. On my Mac the ipch file range from 10-200MB. I compressed a 185MB ipch down to just 21MB using zstd.

I think compression could save vscode users a lot of space (especially MacBook owners with 128/256GB storage). If CPU usage and battery life become a concern, maybe add it as an option? Maybe another option to adjust the compression ratio?
